biogeochem/FatesLandUseChangeMod.F90

```diff
@@ -0,0 +1,307 @@
+    ! aggregate land use types to those being used in the simulation, and output a transition matrix that can be used to drive patch
+    ! disturbance rates.
+
+    ! !ARGUMENTS:
+    type(bc_in_type) , intent(in) :: bc_in
+    real(r8), intent(inout) :: landuse_transition_matrix(n_landuse_cats, n_landuse_cats)  ! [m2/m2/day]
+
+    ! !LOCAL VARIABLES:
+    type(luh2_fates_lutype_map) :: lumap
+    integer :: i_donor, i_receiver, i_luh2_transitions, i_luh2_states
+    character(5) :: donor_name, receiver_name
+    character(14) :: transition_name
+    real(r8) :: urban_fraction
+    real(r8) :: temp_vector(hlm_num_luh2_transitions)
+    logical  :: modified_flag
+
+    ! zero the transition matrix and the urban fraction
+    landuse_transition_matrix(:,:) = 0._r8
+    urban_fraction = 0._r8
+
+    use_luh_if: if ( hlm_use_luh .eq. itrue ) then
+
+       ! Check the LUH data incoming to see if any of the transitions are NaN
+       temp_vector = bc_in%hlm_luh_transitions
+       call CheckLUHData(temp_vector,modified_flag)
+       if (.not. modified_flag) then
+          ! identify urban fraction so that it can be factored into the land use state output
+          urban_fraction = bc_in%hlm_luh_states(findloc(bc_in%hlm_luh_state_names,'urban',dim=1))
+       end if
+
+       !!TODO: may need some logic here to ask whether or not ot perform land use change on this timestep. current code occurs every day.
+       !!If not doing transition every day, need to update units.
+
+       transitions_loop: do i_luh2_transitions = 1, hlm_num_luh2_transitions
+
+          ! transition names are written in form xxxxx_to_yyyyy where x and y are donor and receiver state names
+          transition_name = bc_in%hlm_luh_transition_names(i_luh2_transitions)
+          donor_name = transition_name(1:5)
+          receiver_name = transition_name(10:14)
+
+          ! Get the fates land use type index associated with the luh2 state types
+          i_donor= lumap%GetIndex(donor_name)
+          i_receiver = lumap%GetIndex(receiver_name)
+
+          ! Avoid transitions with 'urban' as those are handled seperately
+          if (.not.(i_donor .eq. fates_unset_int .or. i_receiver .eq. fates_unset_int)) then
+             landuse_transition_matrix(i_donor,i_receiver) = &
+                  landuse_transition_matrix(i_donor,i_receiver) +  temp_vector(i_luh2_transitions) * years_per_day / (1._r8 - urban_fraction)
+
+          end if
+       end do transitions_loop
+    end if use_luh_if
+  end subroutine get_landuse_transition_rates
+
+  !----------------------------------------------------------------------------------------------------
+
+  function GetLUCategoryFromStateName(this, state_name) result(landuse_category)
+
+    class(luh2_fates_lutype_map) :: this
+    character(len=5), intent(in) :: state_name
+    integer :: landuse_category
+
+    landuse_category = this%landuse_categories(findloc(this%state_names,state_name,dim=1))
+
```

Comment:
If bogus names are passed in, findloc will return a zero:
https://www.intel.com/content/www/us/en/docs/fortran-compiler/developer-guide-reference/2023-1/findloc.html
So it seems impossible that the model will chug along if there is not agreement between fates definitions and the host/dataset definitions of the category names.

Reply:
I'm not going to push hard for this change, but I think ideally we would have a graceful check somewhere that would provide an informative message that the fates and host/dataset landclass names do not match up here.  At least the model will not continue to run using mismatched classes.

Reply:
Thanks @rgknox, this is a good point.  Added.